src/rust/catmem/queue.rs

```diff
@@ -7,66 +7,192 @@
 
 use super::pipe::Pipe;
 use crate::{
-    collections::shared_ring::SharedRingBuffer,
     runtime::{
         fail::Fail,
+        limits,
+        memory::DemiBuffer,
         queue::IoQueue,
         QType,
     },
     scheduler::{
         TaskHandle,
+        Yielder,
         YielderHandle,
     },
 };
-use ::std::collections::HashMap;
+use ::std::{
+    cell::{
+        RefCell,
+        RefMut,
+    },
+    collections::HashMap,
+    rc::Rc,
+};
+
+//======================================================================================================================
+// Constants
+//======================================================================================================================
+
+/// Maximum number of retries for pushing a EoF signal.
+const MAX_RETRIES_PUSH_EOF: u32 = 16;
 
 //======================================================================================================================
 // Structures
 //======================================================================================================================
 
 /// Per-queue metadata: reference to the shared pipe.
+#[derive(Clone)]
 pub struct CatmemQueue {
-    pipe: Pipe,
-    pending_ops: HashMap<TaskHandle, YielderHandle>,
+    pipe: Rc<RefCell<Pipe>>,
+    pending_ops: Rc<RefCell<HashMap<TaskHandle, YielderHandle>>>,
 }
 
 //======================================================================================================================
 // Associated Functions
 //======================================================================================================================
 
 impl CatmemQueue {
-    pub fn new(ring: SharedRingBuffer<u16>) -> Self {
-        Self {
-            pipe: Pipe::new(ring),
-            pending_ops: HashMap::<TaskHandle, YielderHandle>::new(),
+    pub fn new(name: &str) -> Result<Self, Fail> {
```

Comment:
Could we rename this so that it matches the operation on the libos layer? (ex: `create_pipe()`)

Reply:
Not sure what to do about this. Would be ok to changing everything back to pipe instead of ring but it's weird to name it create_pipe when it actually create a shared ring buffer.

Reply:
Sorry, I was reviewing old code.